From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 12 Feb 2024 20:10:08 +0100
Subject: [PATCH] Ignore system directories

Ignore technical directories, such as `.git` for Git repositories or `.DS_Store` on macOS file systems,
when scanning directories, such as when reading a data pack from a plain directory.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/packs/PathPackResources.java b/src/main/java/net/minecraft/server/packs/PathPackResources.java
index 4822f94ce183a99ad9e0d1bdc6c5708d958f6104..c1215d5a809b4c0b50e13cfa38611429ad22408e 100644
--- a/src/main/java/net/minecraft/server/packs/PathPackResources.java
+++ b/src/main/java/net/minecraft/server/packs/PathPackResources.java
@@ -108,6 +108,11 @@ public class PathPackResources extends AbstractPackResources {
         try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
             for(Path path2 : directoryStream) {
                 String string = path2.getFileName().toString();
+                // Fiddle start - ignore system directories
+                if (org.fiddlemc.fiddle.pack.read.SystemFiles.isSystemDirectory(string)) {
+                    continue;
+                }
+                // Fiddle end - ignore system directories
                 // Paper start
                 if (!Files.isDirectory(path2)) {
                     LOGGER.error("Invalid directory entry: {} in {}.", string, this.root, new java.nio.file.NotDirectoryException(string));
diff --git a/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java b/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java
index 36bc04fdb3da81283a1a85f7c649b0ceddd163ab..fcc91d3a24a9b7527890acb5605eeb55c17f930c 100644
--- a/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java
+++ b/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java
@@ -65,6 +65,12 @@ public class FolderRepositorySource implements RepositorySource {
         try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
             for(Path path2 : directoryStream) {
                 try {
+                    // Fiddle start - ignore system directories
+                    var path2Filename = path2.getFileName();
+                    if (path2Filename != null && org.fiddlemc.fiddle.pack.read.SystemFiles.isSystemDirectory(path2Filename.toString())) {
+                        continue;
+                    }
+                    // Fiddle end - ignore system directories
                     List<ForbiddenSymlinkInfo> list = new ArrayList<>();
                     Pack.ResourcesSupplier resourcesSupplier = folderPackDetector.detectPackResources(path2, list);
                     if (!list.isEmpty()) {
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/read/SystemFiles.java b/src/main/java/org/fiddlemc/fiddle/pack/read/SystemFiles.java
new file mode 100644
index 0000000000000000000000000000000000000000..43f1a56419b04aefe75c2e940fe2fe7501fc1468
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/read/SystemFiles.java
@@ -0,0 +1,21 @@
+// Fiddle - ignore system directories
+
+package org.fiddlemc.fiddle.pack.read;
+
+public class SystemFiles {
+
+    private final static String[] DIRECTORIES = {
+        ".git",
+        ".DS_Store",
+    };
+
+    public static boolean isSystemDirectory(String filename) {
+        for (var systemDirectory : DIRECTORIES) {
+            if (systemDirectory.equals(filename)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+}
